shared/route-tree/index.js

```diff
@@ -0,0 +1,229 @@
+// @flow
+import * as I from 'immutable'
+import {Component} from 'react'
+
+import type {ConnectedComponent} from 'react-redux'
+import type {ConnectedComponent as TypedConnectedComponent} from '../util/typed-connect'
+
+type LeafTagsParams = {
+  modal: boolean,
+  underStatusBar: boolean,
+}
+
+export const LeafTags: (spec?: LeafTagsParams) => LeafTagsParams & I.Record<LeafTagsParams> = I.Record({
+  modal: false,
+  underStatusBar: false,
+})
+
+const _RouteDefNode = I.Record({
+  defaultSelected: null,
+  component: null,
+  containerComponent: null,
+  tags: LeafTags(),
+  initialState: I.Map(),
+  children: I.Map(),
+})
+
+type RouteDefParams<P> = {
+  defaultSelected?: string,
+  tags?: LeafTags,
+  initialState?: {},
+  children: {},
+} & (
+  // This lengthy type definition was necessary to get all of our component permutations to type check.
+  { component?: Component<any, P, any> | $Supertype<Component<any, P, any>> | Class<ConnectedComponent<P, any, any, any>> | Class<TypedConnectedComponent<P>> }
+  | { containerComponent: Component<any, P, any> }
+)
+
+export class RouteDefNode extends _RouteDefNode {
+  constructor ({defaultSelected, component, containerComponent, tags, initialState, children}: RouteDefParams<*>) {
+    super({
+      defaultSelected: defaultSelected || null,
+      component,
+      containerComponent,
+      tags: LeafTags(tags),
+      initialState: I.Map(initialState),
+      props: I.Map(),
+      state: I.Map(),
+      children: I.Seq(children)
+        .map(params => params instanceof RouteDefNode || typeof params === 'function' ? params : new RouteDefNode(params))
+        .toMap(),
+    })
+  }
+
+  getChild (name: string): ?RouteDefNode {
+    const childDef = this.children.get(name)
+    if (!childDef) {
+      return
+    }
+    if (typeof childDef === 'function') {
+      return childDef()
+    }
+    return childDef
+  }
+}
+
+type RouteStateParams = {
+  selected: string | null,
+  props?: I.Map<string, any>,
+  state?: I.Map<string, any>,
+}
+
+const _RouteStateNode = I.Record({
+  selected: null,
+  props: I.Map(),
+  state: I.Map(),
+  children: I.Map(),
+})
+
+export class RouteStateNode extends _RouteStateNode {
+  constructor (data: RouteStateParams) {  // eslint-disable-line no-useless-constructor
```

Comment:
Eslint doesn't like this constructor, but in theory it helps provide types for the Record constructor.